File: src/main/java/org/jetbrains/IntJavaStreamBenchmark.java
package org.jetbrains;

import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.infra.Blackhole;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

@State(Scope.Thread)
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.SECONDS)
public class IntJavaStreamBenchmark {
    public static <E> ArrayList<E> makeList(Iterable<E> iter) {
        ArrayList<E> list = new ArrayList<E>();
        for (E item : iter) {
            list.add(item);
        }
        return list;
    }

    public ArrayList<Integer> data = makeList(JetbrainsPackage.intValues());

    @Benchmark
    public Long filterAndCount() {
        return data.stream().filter(it -> it % 2 == 0).count();
    }

    @Benchmark
    public void filterAndMap(final Blackhole bh) {
        Stream<Integer> stream = data.stream().filter(it -> it % 2 == 0).map(it -> it * 10);
        stream.forEach(it -> bh.consume(it));
    }

    @Benchmark
    public Long countFiltered() {
        return data.stream().mapToLong(it -> it % 2 == 0 ? 1 : 0).sum();
    }
}

File: src/main/kotlin/org/jetbrains/IntJavaStreamBenchmark.kt
package org.jetbrains

import org.openjdk.jmh.annotations.*
import java.util.concurrent.*
import java.util.stream.*
import org.openjdk.jmh.infra.Blackhole
import java.util.ArrayList

State(Scope.Thread)
BenchmarkMode(Mode.Throughput)
OutputTimeUnit(TimeUnit.SECONDS)
open class IntJavaStreamBenchmark {
    val list: ArrayList<Int> = ArrayList(intValues().toList())
    val getStreamMethod = list.javaClass.getMethod("stream")
    val data: Stream<Int>
        get() = getStreamMethod.invoke(list) as Stream<Int>

    Benchmark fun filterAndCount(bh: Blackhole) {
        val stream = data.filter(predicate<Int>({ it % 2 == 0 }))!!
        bh.consume(stream.count())
    }

    Benchmark fun countFiltered(bh: Blackhole) {
        bh.consume(data.mapToLong(longFunction({ if (it % 2 == 0) 1L else 0L }))!!.sum())
    }

    Benchmark fun countFilteredWithValue(bh: Blackhole) {
        val value = data.mapToLong(longFunction({ if (it % 2 == 0) 1L else 0L }))!!.sum()
        bh.consume(value)
    }

    Benchmark fun filterAndMap(bh: Blackhole) {
        val stream = data.filter(predicate<Int>({ it % 2 == 0 }))!!.map(function<Int, Int>({ it * 10 }))!!
        stream.forEach(consume({ bh.consume(it) }))
    }
}
